examples/vision/image_captioning.py

```diff
@@ -0,0 +1,509 @@
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = layers.Dense(embed_dim, activation="relu")
+        self.layernorm_1 = layers.LayerNormalization()
+
+    def call(self, inputs, training, mask=None):
+        inputs = self.dense_proj(inputs)
+        attention_output = self.attention(
+            query=inputs, value=inputs, key=inputs, attention_mask=None
+        )
+        proj_input = self.layernorm_1(inputs + attention_output)
+        return proj_input
+
+
+class PositionalEmbedding(layers.Layer):
+    def __init__(self, sequence_length, vocab_size, embed_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.token_embeddings = layers.Embedding(
+            input_dim=vocab_size, output_dim=embed_dim
+        )
+        self.position_embeddings = layers.Embedding(
+            input_dim=sequence_length, output_dim=embed_dim
+        )
+        self.sequence_length = sequence_length
+        self.vocab_size = vocab_size
+        self.embed_dim = embed_dim
+
+    def call(self, inputs):
+        length = tf.shape(inputs)[-1]
+        positions = tf.range(start=0, limit=length, delta=1)
+        embedded_tokens = self.token_embeddings(inputs)
+        embedded_positions = self.position_embeddings(positions)
+        return embedded_tokens + embedded_positions
+
+    def compute_mask(self, inputs, mask=None):
+        return tf.math.not_equal(inputs, 0)
+
+
+class TransformerDecoderBlock(layers.Layer):
+    def __init__(self, embed_dim, ff_dim, num_heads, **kwargs):
+        super().__init__(**kwargs)
+        self.embed_dim = embed_dim
+        self.ff_dim = ff_dim
+        self.num_heads = num_heads
+        self.attention_1 = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.attention_2 = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = keras.Sequential(
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim)]
+        )
+        self.layernorm_1 = layers.LayerNormalization()
+        self.layernorm_2 = layers.LayerNormalization()
+        self.layernorm_3 = layers.LayerNormalization()
+
+        self.pe = PositionalEmbedding(
+            embed_dim=EMBED_DIM, sequence_length=SEQ_LENGTH, vocab_size=VOCAB_SIZE
+        )
+        self.out = layers.Dense(VOCAB_SIZE, activation="softmax")
+        self.dropout = layers.Dropout(0.5)
+
+    def call(self, inputs, encoder_outputs, training, mask=None):
+        inputs = self.pe(inputs)
+        causal_mask = self.get_causal_attention_mask(inputs)
+
+        attention_output_1 = self.attention_1(
+            query=inputs, value=inputs, key=inputs, attention_mask=causal_mask
+        )
+        out_1 = self.layernorm_1(inputs + attention_output_1)
+
+        attention_output_2 = self.attention_2(
+            query=out_1, value=encoder_outputs, key=encoder_outputs, attention_mask=None
```

Comment:
You need to provide the padding mask here, no?

Chances are this will improve your results. Sorry for missing this earlier.

Reply:
No worries Francois. I will try adding it